zkvm/src/signature/signer.rs

```diff
@@ -165,28 +155,22 @@ impl PartyAwaitingShares {
         Signature { s, R: self.R }
     }
 
-    pub fn receive_shares(
-        self,
-        shares: Vec<Scalar>,
-        pubkeys: Vec<VerificationKey>,
-    ) -> Result<Signature, VMError> {
-        // Check that all shares are valid
-        for (i, s_i) in shares.iter().enumerate() {
-            let S_i = s_i * RISTRETTO_BASEPOINT_POINT;
-            let X_i = pubkeys[i];
-            let R_i = self.nonce_commitments[i].0;
-
-            // Make a_i = H(L, X_i)
-            let a_i = self.multikey.factor_for_key(&VerificationKey(X_i.0));
-
-            let X_i = X_i.0.decompress().ok_or(VMError::InvalidPoint)?;
+    pub fn receive_shares(self, shares: Vec<Scalar>) -> Result<Signature, VMError> {
+        // Create local copies of challenge and &multikey,
+        // since iterator takes ownership of `self` later.
+        let challenge = self.c;
+        let multikey = &self.multikey;
 
-            // Check that S_i = R_i + c * a_i * X_i
-            if S_i != R_i + self.c * a_i * X_i {
-                return Err(VMError::ShareError { index: i });
-            }
-        }
+        // Check that all shares are valid
+        let validated_shares = self
+            .counterparties
+            .into_iter()
+            .zip(shares)
+            .map(|(counterparty, share)| counterparty.sign(share, challenge, multikey))
+            .collect::<Result<Vec<_>, _>>()?;
```

Comment:
wouldn't this work directly, without allocating another Vec?
```suggestion
            .sum()?;
```

Reply:
Done!